### PR TITLE
vstreams cell alias fallback v13

### PR DESCRIPTION
**Pull request by pbibra**
## Description

**This is just a temporary solution internally until we roll out a full TabletPicker refactor as per this issue:** https://github.com/vitessio/vitess/issues/11999

Allow for cell alias fallback during tablet selection for VStreams when client does not specify list of cells. In addition, add the option for local cell preference during tablet selection.

Overall, this PR is trying to solve 2 problems:

Problem 1: We're trying to avoid passing in the list of cells/cell alias that TabletPicker can choose from in the gRPC request itself. Instead, we will now select a candidate from a list of tablets in the VTGate's local cell + any cell within a defined cell alias for the local cell, if one exists.

e.g. We have a cell alias called region-1 which includes the following cells {region-1a, region-1b, region-1c}
A VTGate in region-1b sets up a VStream. The gRPC VStreamRequest object does not have any cells sent from the client, but a cell alias is defined.

As a result TabletPicker.PickForStreaming() will choose from candidates not just in the VTGate's local cell, but in all cells within the region-1 alias without the client having to explicitly specify this.

Problem 2: Of all the candidates selected by TabletPicker, we need a way to allow it to give priority to the local cell. This is where the localPreferenceHint comes in. Two paths in which this can be specified:

a. If no cells specified by the client, send in the list of cells to the TabletPicker as local:region-1b, region-1. The first item in the list is the local cell we want to prioritize and the second is the cell alias that region-1b belongs to.

b. The second way is to allow the client side to also specify this if they choose to pass in the cells within the gRPC request itself rather than use the new flag. They'd send in local,region-1 or if no cell alias, then something like local:,region-1a, region-1c within the VStreamRequest object.

In both cases, TabletPicker can then order candidates like:

tablet-1-region-1b - PRIMARY
tablet-2-region-1b - REPLICA
tablet-1-region-1c - REPLICA
tablet-1-region-1a - REPLICA

where tablet-1-region-1b would be prioritized since it is in the local cell. The in_order hint is then applied on top of this, where we order by tablet type within each "group" (local cells first, then all others). For example, if the tablet type ordering is inOrder:REPLICA,PRIMARY, then we get the following ordering of candidates:

tablet-2-region-1b - REPLICA
tablet-1-region-1b - PRIMARY
tablet-1-region-1c - REPLICA
tablet-1-region-1a - REPLICA

## Related Issue(s)

https://github.com/vitessio/vitess/issues/11999

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Review by HenryCaiHaiying — APPROVED** [review]
LGTM

File: go/vt/vtgate/vstream_manager.go [hunk 1]
```diff
@@ -411,18 +411,51 @@ func (vs *vstream) alignStreams(ctx context.Context, event *binlogdatapb.VEvent,
...
+			// send in the vtgate's cell for local cell preference
+			cells = append(cells, fmt.Sprintf("local:%s", vs.vsm.cell))
+		}
+		cells = append(cells, alias...)
+	}
+
 	if len(cells) == 0 {
 		// use the vtgate's cell by default
 		cells = append(cells, vs.vsm.cell)
 	}
```
Comment (tanjinx) [inline go/vt/vtgate/vstream_manager.go:456]: should we remove these 4 lines?

**Review by tanjinx — APPROVED** [review]

File: go/vt/vtgate/vstream_manager.go [hunk 2]
```diff
@@ -411,18 +411,46 @@ func (vs *vstream) alignStreams(ctx context.Context, event *binlogdatapb.VEvent,
...
 	if len(cells) == 0 {
-		// use the vtgate's cell by default
-		cells = append(cells, vs.vsm.cell)
+		log.Info("[VSTREAM MANAGER] no cells specified by client, falling back to alias...")
+		// append the alias this cell belongs to, otherwise appends the vtgate's cell
+		alias := topo.GetAliasByCell(ctx, vs.ts, vs.vsm.cell)
+		// an alias was actually found
+		if alias != vs.vsm.cell {
+			// send in the vtgate's cell for local cell preference
+			cells = append(cells, fmt.Sprintf("local:%s", vs.vsm.cell))
```
Comment (brirams) [inline go/vt/vtgate/vstream_manager.go:448]: it's unclear why we need to lookup the cell alias if the tablet picker already does cell alias lookup here: https://github.com/slackhq/vitess/blob/5563eb126e91457ef0dde6e51f6375efd0532a01/go/vt/discovery/tablet_picker.go#L195-L200

**Review by brirams — APPROVED** [review]
i think if we were going to do this again and we wanted to do something generalizable, i would've decomposed the `TabletPicker` struct a bit to take in some functors that implement different ways of sorting and picking cells, as well as sorting and picking tablets. I think encoding logic into special strings in arguments is a bit of a code smell but there is certainly plenty of precedent for that and, especially if it does what we need it to do, i say go for it!  thank you for including so many tests!

File: go/vt/vtgate/vstream_manager.go (hunk 2 above)
Comment (pbibra) [reply go/vt/vtgate/vstream_manager.go:448]: It only does a cell lookup if the string passed into "tp.cells" is not a valid cell: https://github.com/slackhq/vitess/blob/5563eb126e91457ef0dde6e51f6375efd0532a01/go/vt/discovery/tablet_picker.go#L191

**Comment by pbibra** [comment]
> i think if we were going to do this again and we wanted to do something generalizable, i would've decomposed the `TabletPicker` struct a bit to take in some functors that implement different ways of sorting and picking cells, as well as sorting and picking tablets. I think encoding logic into special strings in arguments is a bit of a code smell but there is certainly plenty of precedent for that and, especially if it does what we need it to do, i say go for it! thank you for including so many tests!

I totally agree! That's actually the longer term plan for TabletPicker that I'm working on with PlanetScale. This is a temporary patch to unblock us internally. The hope is that once the long term changes are in upstream, we can adopt them as well and get rid of this. https://github.com/vitessio/vitess/issues/11999

**Pull request merged**